### PR TITLE
EuroSciPy 2025: Predictive modeling with imbalanced datasets

### PR DESCRIPTION
**Pull request by glemaitre**

### CODE REVIEW

File: content/python_files/imbalance_correction.py [hunk 1]
```diff
@@ -0,0 +1,853 @@
...
+negative_indices = rng.choice(
+    np.arange(len(y_past))[y_past == 0],
+    size=3 * len(y_positive),
+    replace=False,
+)
+
+X_observed = pd.concat([X_positive, X_past.iloc[negative_indices]])
+y_observed = pd.concat([y_positive, y_past.iloc[negative_indices]])
+
+X_observed.shape
```
Comment (glemaitre) [inline content/python_files/imbalance_correction.py]:
```suggestion
mask_positive_target = y_past == 1
X_positive = X_past[mask_positive_target].copy()
y_positive = y_past[mask_positive_target].copy()

rng = np.random.default_rng(0)
negative_indices = rng.choice(
    np.flatnonzero(~mask_positive_target),
    size=3 * len(y_positive),
    replace=False,
)

X_observed = pd.concat([X_positive, X_past.iloc[negative_indices]])
y_observed = pd.concat([y_positive, y_past.iloc[negative_indices]])

X_observed.shape
```

Comment (glemaitre) [reply content/python_files/imbalance_correction.py]: Just storing the `y_past == 1`

Comment (glemaitre) [reply content/python_files/imbalance_correction.py]: I also used `np.flatnonzero` instead of `arange` + indexing with mask.

File: content/python_files/imbalance_correction.py [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+# %%
+from sklearn.model_selection import train_test_split
+
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X_observed, y_observed, test_size=0.5, random_state=0
+)
+
+# %%
+logreg_params = dict(penalty=None, tol=1e-8)
```
Comment (glemaitre) [inline content/python_files/imbalance_correction.py]: Do we want to use this for the entire notebooks? I see a couple of occurrences where we don't set the tolerance?

File: content/python_files/imbalance_correction.py [hunk 3]
```diff
@@ -0,0 +1,853 @@
...
+# intercept.
+
+# %%
+from sklearn.linear_model import LogisticRegression
+
+cheating_model = LogisticRegression(penalty=None).fit(X_future, y_future)
+
+
+# %%
+class ModelComparator:
```
Comment (glemaitre) [inline content/python_files/imbalance_correction.py]: It is a nice class. I'm wondering if we should have one single documentation string for each method?

**Review by glemaitre — COMMENTED** [review]
@ogrisel I made a pass on the notebook and it looks all good. I would not mind to merge it quickly to be able to check the rendering on the website and make new PR if we spot something wrong (notably for the equations :) )

File: content/python_files/imbalance_correction.py (hunk 3 above)
Comment (ogrisel) [reply content/python_files/imbalance_correction.py]: I'll add docstring later.

File: content/python_files/imbalance_correction.py (hunk 2 above)
Comment (ogrisel) [reply content/python_files/imbalance_correction.py]: We could, but initially, I used it to explore the impact of using a non-zero penalty on all the models trained on the small subsampled data to debug other things.
